model/speedscope_profile.go

```diff
@@ -38,41 +40,373 @@ type SpeedscopeProfileEvent struct {
-	At    float64 `json:"at"`    // XXX can this be int64?
-	Frame int64   `json:"frame"` // Index into frames array.
+type EventedProfile struct {
+	EndValue float64 `json:"endValue"`
+	Events []FrameEvent `json:"events"`
+	Name string `json:"name"`
+	StartValue float64 `json:"startValue"`
+	Type ProfileType `json:"type"`
+	Unit ValueUnit `json:"unit"`
 }
 
-// XXX How to handle switch on EventedProfile | SampledProfile? Currently just handling EventedProfile
-type SpeedscopeProfile struct {
-	Type       string                 `json:"type"` // XXX enum 'evented' or 'sampled'
-	Name       string                 `json:"name"`
-	Unit       string                 `json:"unit"`       // XXX enum
-	StartValue float64                `json:"startValue"` // XXX not sure if can be int64
-	EndValue   float64                `json:"endValue"`   // XXX not sure if can be int64
-	Events     []SpeedscopeFrameEvent `json:"events"`
+func (e EventedProfile) isProfile() {}
+
+type SampledProfile struct {
+	EndValue float64 `json:"endValue"`
+	Name string `json:"name"`
+	Samples [][]float64 `json:"samples"`
+	StartValue float64 `json:"startValue"`
+	Type ProfileType `json:"type"`
+	Unit ValueUnit `json:"unit"`
+	Weights []float64 `json:"weights"`
+}
+
+func (e SampledProfile) isProfile() {}
+
+type FrameEvent struct {
+	At float64 `json:"at"`
+	Frame int64 `json:"frame"`
+	Type FrameEventType `json:"type"`
 }
-type SpeedscopeFrame struct {
+
+type Shared struct {
+	Frames []Frame `json:"frames"`
+}
+
+type Frame struct {
+	Col *int64 `json:"col,omitempty"`
+	File *string `json:"file,omitempty"`
+	Line *int64 `json:"line,omitempty"`
 	Name string `json:"name"`
-	File string `json:"file"`
-	Line int64  `json:"line"`
-	Col  int64  `json:"col"`
 }
-type SpeedscopeShared struct {
-	Frames []SpeedscopeFrame `json:"frames"`
+
+type FrameEventType string
+
+const Close FrameEventType = "C"
+const Open FrameEventType = "O"
+
+var eventTypes = []interface{}{
+	"C",
+	"O",
 }
-type SpeedscopeFile struct {
-	Schema   string              `json:"$schema"`
-	Shared   SpeedscopeShared    `json:"shared"`
-	Profiles []SpeedscopeProfile `json:"profiles"`
+
+type ProfileType string
+
+const Evented ProfileType = "evented"
+const Sampled ProfileType = "sampled"
+
+var profileTypes = []interface{}{
+	"evented",
+	"sampled",
+}
+
+type ValueUnit string
+
+const None ValueUnit = "none"
+const Bytes ValueUnit = "bytes"
+const Microseconds ValueUnit = "microseconds"
+const Milliseconds ValueUnit = "milliseconds"
+const Nanoseconds ValueUnit = "nanoseconds"
+const Seconds ValueUnit = "seconds"
+
+var valueUnits = []interface{}{
+	"bytes",
+	"microseconds",
+	"milliseconds",
+	"nanoseconds",
+	"none",
+	"seconds",
+}
+
+func (j *ValueUnit) UnmarshalJSON(b []byte) error {
+	var v string
+	if err := json.Unmarshal(b, &v); err != nil {
+		return err
+	}
+	var ok bool
+	for _, expected := range valueUnits {
+		if reflect.DeepEqual(v, expected) {
```

Comment:
I'm curious, was this code generated by that tool you mentioned? I've heard (I have little go experience) that `reflect.*` can be quite slow.

Reply:
Yeah, this was generated by the tool from the JSON schema. It's probably not the nicest or best performing code (hack it til your demo makes it?!)